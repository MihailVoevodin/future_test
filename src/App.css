.App {
  box-sizing: border-box;
  text-align: center;
  color: black;
}

.container {
  max-width: 1170px;
  padding: 0 15px;
  margin: 0 auto;
}

a {
  text-decoration: none;
  color: inherit;
}

ul {
  list-style: none;
}

h1 {
  margin: 0;
  padding-bottom: 20px;
}

header {
  color: white;
  background-image: url("assets/bg.jpg");
  background-repeat: no-repeat;
  background-size: cover;
  padding: 20px 0;
}
