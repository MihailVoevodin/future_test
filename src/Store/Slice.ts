import {createSlice, createAsyncThunk, PayloadAction} from '@reduxjs/toolkit';
import {BooksService} from 'api/BooksService';

export const loadBooksList = createAsyncThunk(
    'books/getBooksList',
    async ({inputValue, startIndex, sorting, filter}: {inputValue: string, startIndex: number, sorting: string, filter: string}, {rejectWithValue, dispatch}) => {
        const response = await BooksService.getBooksList(inputValue, startIndex, sorting, filter);

        if (!response.data.items) {
            dispatch(setIsError())
            dispatch(setIsLoading())
            return;
        }

        if (response.status !== 200) {
            return rejectWithValue('Server Error!');
        }

        return response.data;
    }
);

export const loadPaginateBooksList = createAsyncThunk(
    'books/getPaginateBooksList',
    async ({inputValue, startIndex, sorting, filter}: {inputValue: string, startIndex: number, sorting: string, filter: string}, {rejectWithValue}) => {
        const response = await BooksService.getBooksList(inputValue, startIndex, sorting, filter);

        if (response.status !== 200) {
            return rejectWithValue('Server Error!');
        }

        return response.data;
    }
);

export const loadBook = createAsyncThunk(
    'books/getBook',
    async (id: string | undefined, {rejectWithValue}) => {
        const response = await BooksService.getBook(id);

        if (response.status !== 200) {
            return rejectWithValue('Server Error!');
        }

        return response.data;
    }
);

/**
 * Интерфейс redux-ветки книг.
 * @param inputValue Значение инпута поиска.
 */
export interface IState {
    inputValue: string;
    isLoading: boolean;
    isError: boolean;
    errorMessage: string;
    booksList: IBook[];
    numberOfBooks: number;
    startIndex: number;
    book: Partial<IBook>;
    sorting: string;
    filter: string;
    isPaginateLoad: boolean;
}

const initialState: IState = {
    inputValue: '',
    isLoading: false,
    isError: false,
    errorMessage: '',
    booksList: [],
    numberOfBooks: 0,
    startIndex: 0,
    book: {},
    sorting: 'relevance',
    filter: 'all',
    isPaginateLoad: false,
};

/**
 * Срез книг.
 */
const BooksSlice = createSlice({
    name: 'books',
    initialState: initialState,
    reducers: {
        setInputValue(state, action: PayloadAction<string>) {
            state.inputValue = action.payload;
        },
        setStartIndex(state, action: PayloadAction<number>) {
            state.startIndex = action.payload;
        },
        setBooksList(state) {
            state.booksList = [];
        },
        setBook(state, action) {
            state.book = action.payload;
        },
        setFilter(state, action: PayloadAction<string>) {
            state.filter = action.payload;
        },
        setSorting(state, action: PayloadAction<string>) {
            state.sorting = action.payload;
        },
        setIsError(state) {
            state.isError = true;
            state.errorMessage = 'No matching results';
        },
        setIsLoading(state) {
            state.isLoading = false;
        },
    },
    extraReducers: (builder) => {
        builder
            .addCase(loadBooksList.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(loadBooksList.rejected, (state) => {
                state.isError = true;
                state.errorMessage = 'No matching results';
                state.isLoading = false;
            })
            .addCase(loadBooksList.fulfilled, (state, action) => {
                state.booksList = action.payload.items;
                state.numberOfBooks = action.payload.totalItems;
                state.isLoading = false;
                state.isError = false;
                state.errorMessage = '';
            })
            .addCase(loadPaginateBooksList.pending, (state) => {
                state.isPaginateLoad = true;
            })
            .addCase(loadPaginateBooksList.rejected, (state) => {
                state.isError = true;
                state.errorMessage = 'No matching results';
                state.isPaginateLoad = false;
            })
            .addCase(loadPaginateBooksList.fulfilled, (state, action) => {
                state.booksList.push.apply(state.booksList, action.payload.items);
                state.numberOfBooks = action.payload.totalItems;
                state.isPaginateLoad = false;
                state.isError = false;
                state.errorMessage = '';
            })
            .addCase(loadBook.pending, (state) => {
                state.isLoading = true;
            })
            .addCase(loadBook.rejected, (state) => {
                state.isError = true;
                state.errorMessage = 'No matching results';
                state.isLoading = false;
            })
            .addCase(loadBook.fulfilled, (state, action) => {
                state.book = action.payload;
                state.isLoading = false;
                state.isError = false;
                state.errorMessage = '';
            })
    }
});

export const {
    setInputValue,
    setStartIndex,
    setBooksList,
    setBook,
    setSorting,
    setFilter,
    setIsError,
    setIsLoading,
} = BooksSlice.actions;

export default BooksSlice.reducer;
