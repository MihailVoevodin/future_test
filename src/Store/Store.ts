import {configureStore} from '@reduxjs/toolkit';
import BooksSlice, {IState} from 'Store/Slice'

/**
 * @param books - Ветка книг.
 */
export interface IAppState {
    books: IState;
}

export const store = configureStore<IAppState>({
    reducer: {
        books: BooksSlice,
    },
});

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;