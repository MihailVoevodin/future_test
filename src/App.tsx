import {Book} from 'Components/Book/Book';
import {BooksList} from 'Components/BooksList/BooksList';
import {Filters} from 'Components/Filters/Filters';
import React from 'react';
import {BrowserRouter, Route, Routes} from 'react-router-dom';
import './App.css';
import {Search} from 'Components/Search/Search';

function App() {
    return (
        <BrowserRouter>
            <div className="App">
                <header>
                    <h1 className='title'>Search for books</h1>
                    <Search />
                    <Filters />
                </header>
                <main>
                    <Routes>
                        <Route path='/' element={<BooksList />} />
                        <Route path='/book/:id' element={<Book />} />
                    </Routes>
                </main>
            </div>
        </BrowserRouter>
    );
}

export default App;
