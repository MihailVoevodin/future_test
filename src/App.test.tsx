import React from 'react';
import { render, screen } from '@testing-library/react';
import {Provider} from 'react-redux';
import {store} from 'Store/Store';
import App from './App';

describe('App component', () => {

    test('App renders', () => {
        render(
            <Provider store={store}>
                <App />
            </Provider>
        );

        expect(screen.getByText(/search for books/i)).toBeInTheDocument();
    })

    test('App snapshot', () => {
        const app = render(
            <Provider store={store}>
                <App />
            </Provider>
        );

        expect(app).toMatchSnapshot();
    })
});
