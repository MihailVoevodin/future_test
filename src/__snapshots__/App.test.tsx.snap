// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component App snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="App"
      >
        <header>
          <h1
            class="title"
          >
            Search for books
          </h1>
          <form
            name="Search"
          >
            <input
              class="input"
              placeholder="Find books"
              type="text"
            />
            <button
              class="button"
              type="submit"
            >
              <span
                aria-label="search"
                class="anticon anticon-search"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </span>
            </button>
          </form>
          <div
            class="filtrationWrapper"
          >
            <div>
              <span>
                Categories
              </span>
              <select
                data-testid="select"
                name="categories"
              >
                <option
                  data-testid="select-option"
                  value="all"
                >
                  all
                </option>
                <option
                  data-testid="select-option"
                  value="art"
                >
                  art
                </option>
                <option
                  data-testid="select-option"
                  value="biography"
                >
                  biography
                </option>
                <option
                  data-testid="select-option"
                  value="computers"
                >
                  computers
                </option>
                <option
                  data-testid="select-option"
                  value="history"
                >
                  history
                </option>
                <option
                  data-testid="select-option"
                  value="medical"
                >
                  medical
                </option>
                <option
                  data-testid="select-option"
                  value="poetry"
                >
                  poetry
                </option>
              </select>
            </div>
            <div>
              <span>
                Sorting by
              </span>
              <select
                name="sorting"
              >
                <option
                  value="relevance"
                >
                  relevance
                </option>
                <option
                  value="newest"
                >
                  newest
                </option>
              </select>
            </div>
          </div>
        </header>
        <main>
          <div
            style="padding-top: 100px;"
          >
            <div
              aria-busy="true"
              aria-live="polite"
              class="ant-spin ant-spin-spinning css-dev-only-do-not-override-ph9edi"
            >
              <span
                class="ant-spin-dot ant-spin-dot-spin"
              >
                <i
                  class="ant-spin-dot-item"
                />
                <i
                  class="ant-spin-dot-item"
                />
                <i
                  class="ant-spin-dot-item"
                />
                <i
                  class="ant-spin-dot-item"
                />
              </span>
            </div>
          </div>
        </main>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="App"
    >
      <header>
        <h1
          class="title"
        >
          Search for books
        </h1>
        <form
          name="Search"
        >
          <input
            class="input"
            placeholder="Find books"
            type="text"
          />
          <button
            class="button"
            type="submit"
          >
            <span
              aria-label="search"
              class="anticon anticon-search"
              role="img"
            >
              <svg
                aria-hidden="true"
                data-icon="search"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                />
              </svg>
            </span>
          </button>
        </form>
        <div
          class="filtrationWrapper"
        >
          <div>
            <span>
              Categories
            </span>
            <select
              data-testid="select"
              name="categories"
            >
              <option
                data-testid="select-option"
                value="all"
              >
                all
              </option>
              <option
                data-testid="select-option"
                value="art"
              >
                art
              </option>
              <option
                data-testid="select-option"
                value="biography"
              >
                biography
              </option>
              <option
                data-testid="select-option"
                value="computers"
              >
                computers
              </option>
              <option
                data-testid="select-option"
                value="history"
              >
                history
              </option>
              <option
                data-testid="select-option"
                value="medical"
              >
                medical
              </option>
              <option
                data-testid="select-option"
                value="poetry"
              >
                poetry
              </option>
            </select>
          </div>
          <div>
            <span>
              Sorting by
            </span>
            <select
              name="sorting"
            >
              <option
                value="relevance"
              >
                relevance
              </option>
              <option
                value="newest"
              >
                newest
              </option>
            </select>
          </div>
        </div>
      </header>
      <main>
        <div
          style="padding-top: 100px;"
        >
          <div
            aria-busy="true"
            aria-live="polite"
            class="ant-spin ant-spin-spinning css-dev-only-do-not-override-ph9edi"
          >
            <span
              class="ant-spin-dot ant-spin-dot-spin"
            >
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
            </span>
          </div>
        </div>
      </main>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
